src/txn/txn_rollback_to_stable.c

```diff
@@ -428,6 +498,85 @@ __rollback_row_ondisk_fixup_key(WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW
+    WT_CELL *kcell;
+    WT_CELL_UNPACK_KV *vpack, _vpack;
+    WT_DECL_RET;
+    WT_ITEM buf;
+    WT_PAGE *page;
+    WT_UPDATE *upd;
+    char ts_string[5][WT_TS_INT_STRING_SIZE];
+    bool prepared;
+
+    page = ref->page;
+    vpack = &_vpack;
+    WT_CLEAR(buf);
+    upd = NULL;
+
+    kcell = WT_COL_PTR(page, cip);
+    __wt_cell_unpack_kv(session, page->dsk, kcell, vpack);
+    prepared = vpack->tw.prepare;
+
+    if (vpack->tw.durable_start_ts > rollback_timestamp ||
+      (!WT_TIME_WINDOW_HAS_STOP(&vpack->tw) && prepared)) {
+        __wt_verbose(session, WT_VERB_RECOVERY_RTS(session),
+          "on-disk update aborted with start durable timestamp: %s, commit timestamp: %s, "
+          "prepared: %s and stable timestamp: %s",
+          __wt_timestamp_to_string(vpack->tw.durable_start_ts, ts_string[0]),
+          __wt_timestamp_to_string(vpack->tw.start_ts, ts_string[1]), prepared ? "true" : "false",
+          __wt_timestamp_to_string(rollback_timestamp, ts_string[2]));
+        if (!F_ISSET(S2C(session), WT_CONN_IN_MEMORY))
+            /* Allocate tombstone and calls function add update to head of insert list. */
+            return (
+              __rollback_col_ondisk_fixup_key(session, ref, cip, rollback_timestamp, true, recno));
+        else {
+            /*
+             * In-memory database does not have a history store to provide a stable update, so
+             * remove the key.
+             */
+            WT_RET(__wt_upd_alloc_tombstone(session, &upd, NULL));
+            WT_STAT_CONN_DATA_INCR(session, txn_rts_keys_removed);
+        }
+    } else if (WT_TIME_WINDOW_HAS_STOP(&vpack->tw) &&
+      (vpack->tw.durable_stop_ts > rollback_timestamp || prepared)) {
+        /*
+         * Clear the remove operation from the key by inserting the original on-disk value as a
+         * standard update.
+         */
+        WT_RET(__wt_page_cell_data_ref(session, page, vpack, &buf));
+
+        WT_ERR(__wt_upd_alloc(session, &buf, WT_UPDATE_STANDARD, &upd, NULL));
+        upd->txnid = vpack->tw.start_txn;
+        upd->durable_ts = vpack->tw.durable_start_ts;
+        upd->start_ts = vpack->tw.start_ts;
+        F_SET(upd, WT_UPDATE_RESTORED_FROM_DS);
+        WT_STAT_CONN_DATA_INCR(session, txn_rts_keys_restored);
+        __wt_verbose(session, WT_VERB_RECOVERY_RTS(session),
+          "key restored with commit timestamp: %s, durable timestamp: %s txnid: %" PRIu64
+          "and removed commit timestamp: %s, durable timestamp: %s, txnid: %" PRIu64
+          ", prepared: %s",
+          __wt_timestamp_to_string(upd->start_ts, ts_string[0]),
+          __wt_timestamp_to_string(upd->durable_ts, ts_string[1]), upd->txnid,
+          __wt_timestamp_to_string(vpack->tw.stop_ts, ts_string[2]),
+          __wt_timestamp_to_string(vpack->tw.durable_stop_ts, ts_string[3]), vpack->tw.stop_txn,
+          prepared ? "true" : "false");
+    } else
+        /* Stable version according to the timestamp. */
+        return (0);
+
```

Comment:
The current ticket is not fully cover this function, as this is getting updated in the next ticket changes, but FYI, we need to call `__rollback_col_add_update()` here to add the update that is prepared here, and also don't free the upd in success scenario. You can do that by either setting `upd = NULL` after success or add if (0) case and point `err` there. This may be the reason for the failure of the `in-memory` scenario that you mentioned some time back.

Reply:
Ah okay, should that be dealt with in the next ticket then?

Reply:
@Sean04 is this resolved or tbd?